new feature: change behavior of sidebar channel buttons to toggle.

neutralparadoxdev opened the pull request:
clicking on a channel in the sidebar with an open pane closes the pane. Sidebar channel buttons are now effectively toggles.

#466

andymandias commented:
I like the idea here, but I think it would be best to make it a configuration option.  I know some users of the `new-pane` and `replace-pane` settings for `default_action` would prefer not to have this behavior, so I suggest we add a `toggle-pane` setting that will cover this use case.

I'd also like to see if we can address #451 in this PR, which is similar but not exactly the same.  If the two ideas can be merged that'd be ideal, but if they're truly incompatible we could add a fourth `default_action`.

neutralparadoxdev commented:
- Added the associated config option.
- Moved hacked Open code to its own event sequence.
- Added documentation for the config option.\
- Did not tackle #451 as I do not know the required behavior to meet it.
- I carelessly merged...

andymandias reviewed (changes requested):
Sorry bitrie, I forgot to submit this earlier.

andymandias commented:
I removed the `pane_toggling` config since your update to `DefaultAction` means `default_action = "toggle-pane"` will set the configuration.  Otherwise I just reorganized it a bit.

Looks good to me, but let's see if we need to implement a "focus before close" setting.

casperstorm commented:
I think this is confusing.
1. `new-pane` opens a new pane.
2. `replace-pane` opens in the same pane.
3. `toggle-pane` opens in the new pane like `new-pane` but changes the behavior of focus.

I think we should consider refactoring a bit here now we have the chance.
Also because `default_action` is a terrible word (that is on me!).

What if we instead split it up:
```
[sidebar]
open_buffer = "new-pane" | "replace-pane"
focus_buffer = "close"
```

`open_buffer`: use serde rename to ensure backwards compatibility with `default_action`
`focus_buffer`: would be `Option<FocusBuffer>` which defaults to `None`.

Suggestions? Better names?

casperstorm reviewed (changes requested):
I think this is confusing.
1. `new-pane` opens a new pane.
2. `replace-pane` opens in the same pane.
3. `toggle-pane` opens in the new pane like `new-pane` but changes the behavior of focus.

I think we should consider refactoring a bit here now we have the chance.
Also because `default_action` is a terrible word (that is on me!).

What if we instead split it up:
```
[sidebar]
open_buffer = "new-pane" | "replace-pane"
focus_buffer = "close"
```

`open_buffer`: use serde rename to ensure backwards compatibility with `default_action`
`focus_buffer`: would be `Option<FocusBuffer>` which defaults to `None`.

I also think that no matter what we should first FOCUS the pane and then consider the action (`focus_buffer`).

Suggestions? Better names?

andymandias commented:
If I understand correctly, the first setting would describe the type of open action made when clicking, and the second would describe the action made when clicking on a buffer with the focused state. Since they describe slightly different concepts, maybe they should have distinct styling. Something like, `open_action` and `focused_buffer`? Or `focused_buffer_action`?

casperstorm commented:
> If I understand correctly, the first setting would describe the type of open action made when clicking, and the second would describe the action made when clicking on a buffer with the focused state. Since they describe slightly different concepts, maybe they should have distinct styling. Something like, `open_action` and `focused_buffer`? Or `focused_buffer_action`?

Yes! What about:
`buffer_action` and `focused_buffer_action` to keep it similar?

neutralparadoxdev commented:
thanks accepting my suggestion and the help. 

I completely missed the ability to minimize changes by reusing messages regarding opening and close buffers.

andymandias commented:
> Yes! What about: `buffer_action` and `focused_buffer_action` to keep it similar?

Sounds good to me!  Did you want to take a crack at the refactor @neutralparadoxdev?  Happy to help, of course, but it's all yours if you want.

casperstorm commented:
@neutralparadoxdev,
I rebased and updated this PR to make it ready for review. Here's what I did:

* Renamed `default_action` to `buffer_action`.
* Added a `buffer_focused_action` which is a optional enum which has `close-pane` variant. By default this is not set.
  * When you set `buffer_focused_action = "close-pane"` it will close a focused buffer, when pressed.

casperstorm commented:
@andymandias take a look if this make sense to you - I kinda like this new setting.

andymandias reviewed (approved):
LGTM!

casperstorm commented:
Sorry, this was a mistake. I'll reopen.
